### PR TITLE
Improve Dev Mode Indicators

### LINKED ISSUE #7498: More prominent visual indication of how to turn off test mode
### Description

When using the new onboarding, a merchant has an option to quickly set up a test account to test WooPayments without needing to provide live Stripe details. Although we include a prompt for such users to set up live payments when they are ready, it is halfway through the `Overview` page meaning it's not very prominent. We've had reports of users not being able to find the option, and the general expectation is that it should appear in the Settings page.

I'd like to get input from others on how this should work, but my suggestion would be a `Notice` at the top of both the `Overview` and `Settings` pages informing the user that `WooPayments` is in test mode and prompting them to set up live payments when they are ready.

### Acceptance criteria

* When a user has onboarded in test/dev mode, they should have a more prominent message helping them to set up live payments when ready.

### Additional context

- Slack thread with reports from HE's: p1697481389856479-slack-C03KTTK2YMA

### PR DESCRIPTION
**Pull request by dmallory42**
Fixes #7498 

#### Changes proposed in this Pull Request

* Updates the style of the Dev Mode indicators on the Settings page:
* Update to use the "Enable live payments" modal previously only used on the Overview page when `Set up payments` is clicked:
* Removed the "Setup Live Payments" on the Overview page since this will now be handled by the dev mode indicators on the Overview and settings pages.
* Update the style of the `TestModeNotice` component, using the new `InlineBanner` component.

Screenshots:

Overview page in Dev Mode:

Screenshot 2023-11-24 at 14 06 08

Overview page in test mode:

Screenshot 2023-11-24 at 14 10 24

Settings page in test mode:

Screenshot 2023-11-24 at 10 32 51

Settings page in Dev mode:

Screenshot 2023-11-24 at 10 31 47

Modal on Settings Page:

Screenshot 2023-11-23 at 10 59 08

Transactions page:

Screenshot 2023-11-23 at 11 03 08

Transactions details page:

Screenshot 2023-11-23 at 11 02 19

Disputes page:

Screenshot 2023-11-23 at 11 03 12

Disputes details page: 

Screenshot 2023-11-23 at 11 04 44

Deposits page:

Screenshot 2023-11-23 at 11 03 45

#### Testing instructions

* Set up a new JN site with this branch, or locally. Enable Progressive Onboarding (new KYC) in the dev tools, but disable dev mode.
* Onboard a new account in builder mode.  
* Verify that the Overview, Settings and Deposits/Disputes/Transactions pages match the designs - verify the `Learn more` links work correctly.
* Make a test transaction and visit the transaction details page and ensure the test mode notice appears on this page and matches the screenshot above.
* Verify the `Set up payments` button on the Overview page opens the "Setup Live Payments" modal.
* Verify the `Set up payments` button on the Settings page opens the "Setup live payments" modal.
* Use the modal to initiate a full onboarding. This time, onboard a live account.
* After the setup is complete, verify you see a different variant of the test mode notice on the overview page ("WooPayments is in test mode. All transactions will be simulated. Learn more")
* Verify that you can toggle test mode on/off on the settings page, and make a test mode transaction and verify it works.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

Update the docs here: https://woo.com/document/woopayments/testing-and-troubleshooting/dev-mode/?quid=e81abdcf1257f23988197cc4883d4dde

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7768` or branch name `dev/7498-improve-dev-mode` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/7498-improve-dev-mode) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 07ece6bdc6e0047a88c36962f4c4b12c9cd38b52
- Build time: 2023-11-29 20:21:40 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by daquinons** [comment]
Hi @dmallory42! Thanks a lot for working on this 💪 

Haven't reviewed the code, but I have a couple of quick scope comments (well, one and a half, the second one is just a visual nitpick :D ) 

- According to the agreements with Design, we were going to hide the "Test mode" section in Settings (including the title and the blocked toggle). Could you modify it so it's not shown in Dev mode accounts? 🙏 
image

- The nitpick is on the "Learn more" link of the Overview page banner. Could you make it align horizontally with the text of "Set up payments"? I imagine the component doesn't have more horizontal space in the button section and it's compressing the "Learn more" text.

**Comment by dmallory42** [comment]
Thanks David!

> According to the agreements with Design, we were going to hide the "Test mode" section in Settings (including the title and the blocked toggle). Could you modify it so it's not shown in Dev mode accounts? 🙏

Ah, I see that now! I was going off the Figma which is still using a slightly older design where the test mode toggle is present. I'll remove it, thanks for spotting!

> The nitpick is on the "Learn more" link of the Overview page banner. Could you make it align horizontally with the text of "Set up payments"? I imagine the component doesn't have more horizontal space in the button section and it's compressing the "Learn more" text.

Ah yes, I wanted to take care of that but I forgot about it - will address it, thanks! 🙇

File: client/components/test-mode-notice/index.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+interface Props {
+	currentPage: CurrentPage;
+	actions?: ReadonlyArray< {
+		label: string;
+		className?: string;
+		variant?: Button.Props[ 'variant' ];
+		url?: string;
+		onClick?: React.MouseEventHandler< HTMLAnchorElement >;
+	} >;
```
Comment (ismaeldcom) [inline client/components/test-mode-notice/index.tsx]: You can also reuse component type definitions with something like:
```tsx
React.ComponentProps< typeof BannerNotice >[ 'actions' ]
```

File: client/components/test-mode-notice/index.tsx [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	actions,
+	isDetailsView = false,
+	isDevMode = false,
+} ) => {
+	if ( ! isInTestMode() ) return null;
+
+	return (
+		<BannerNotice
+			status="warning"
+			icon={ false }
```
Comment (ismaeldcom) [inline client/components/test-mode-notice/index.tsx]: There is no need to set the icon to false, it behaves like that by default 🙂

**Review by ismaeldcom — APPROVED** [review]
Thanks for taking care of it and going the extra mile by refactoring the whole TestNotice component, I love the new approach relying on union types 🙂

Apart from the mentioned visual details, I've only found a couple of nitpicks in the code. It works as expected, and tests good, so I'm preapproving it.

Let's 🚢  it!

File: client/components/test-mode-notice/index.tsx (hunk 1 above)
Comment (dmallory42) [reply client/components/test-mode-notice/index.tsx]: Great tip! Updated in 02a03fe71bc306b0143ccb0bf19f43cfe77deb1c

File: client/components/test-mode-notice/index.tsx (hunk 2 above)
Comment (dmallory42) [reply client/components/test-mode-notice/index.tsx]: Good to know! Updated in 890dd09f5.

**Comment by dmallory42** [comment]
Hi @bradyv! Can you take a look at these changes and let me know if anything is missed here? Thanks!

For additional context, we have an additional https://github.com/Automattic/woocommerce-payments/issues/7733 to handle the other changes like moving towards using Inline Notices in some places, because I wanted the main focus of this issue to be the Dev Mode Indicator changes. We also have another issue to address the changes on the Woo Home page (to the task list). 🙂

**Comment by bradyv** [comment]
@dmallory42 Just did a test run through and I think the changes look great, just a couple of notes to address before merging.

Clicking the `Continue setup` button from the warning modal brought me to a blank page. When I hit the browser back button I was reverted to the WooPayments state pending setup – which is correct – but that blank page would be an issue. This is the url that came up after clicking the button: `/wp-admin/admin.php?wcpay-connect=1&_wpnonce=8ff52f4a0b&wcpay-disable-onboarding-test-mode=true`

Also when getting to that state and clicking the standard WooPayments "Finish setup" prompt it's still sending me into Stripe's test mode, which feels like a really confusing loop for merchants to end up in. I can't actually find the way to get into live setup.

Here's a map of the interactions I followed

map


---

The second note is minor, but in that warning modal I think we should change `payouts` to `deposits` to remain consistent with our platform.
Screenshot 2023-11-27 at 11 06 35

**Comment by dmallory42** [comment]
Thanks for checking, @bradyv! That's definitely not the intentional experience, the merchant should be redirected back to the WooPayments KYC (our Onboarding flow) rather than seeing that blank page. Just to confirm, did you spin up a JN site to test?

I'll double check what's going on tomorrow and resolve it and your other comments before merging. Thanks!

**Comment by bradyv** [comment]
@dmallory42 yes it was done on JN. I can give it a run through on my usual test site for you, but I'll need a refresher on how to get the test branch installed there

**Comment by dmallory42** [comment]
Thanks! I'd say hold out on doing that local testing until I've had a look, possibly there's some issue related to the various pre-release moving parts so I don't want to take up your time unnecessarily! I'll drop a note in here when it's ready for another look.

**Comment by daquinons** [comment]
Hi @dmallory42! I just tested it locally and in my case, it redirected me to a Sorry... page in the URL http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fonboarding
image

After refreshing, I'm able to see the Payments connect screen (and if I click on "Finish setup", I see the new onboarding flow). Any idea of what could be failing here in this redirection @oaratovskyi ? Last time I tested the deletion screen a couple of weeks ago it was working fine so I'm wondering if this could have been caused by the latest fixes in the redirection logic

**Comment by daquinons** [comment]
I've tried a second time and it happened the same. I wonder if it's somehow because the cache might not be cleared on time so when I get redirected the account still exists in the client?

**Comment by dmallory42** [comment]
Hi @bradyv / @daquinons - The issues you mentioned should be resolved, thanks for spotting them! There was a tricky edge case around the builder flow when Dev mode isn't enabled. 

Could you please test again for a final check? You should be able to spin up a new JN site with the latest version of this branch. 

@oaratovskyi already tested, but just to confirm it's all good! 

Thanks!

**Comment by bradyv** [comment]
@dmallory42 When selecting `I'm setting up a store for a client` in the profiler I get an error when trying to install WooPayments:

Screenshot 2023-11-28 at 10 39 47

This happened with two different JN configurations
1. I selected WooCommerce, WooPayments, and this branch from the beta tool
2. I selected WooCommerce, this branch from the beta tool, but _did not_ select WooPayments

This is likely just a JN issue but I wanted to flag it anyway just in case there's something in there to be fixed.

If I skip the store profiler at the point of error I arrive at the Woo home, and can click through to WooPayments and can go through the setup to get into dev mode without any issues. Once in dev mode I see our new prompts and they click through to the first step of Payments Connect without any issues.

**Comment by daquinons** [comment]
@dmallory42 Thanks for looking into the issue, I see how tricky that was here! p1701183918850489-slack-C03KTTK2YMA

- I've disabled Dev Mode in the Dev Tool and chose the builder mode. The deletion of the account works perfectly, and I apologize because last time I was trying with Dev mode enabled and I know that created a lot of issues! I'm happy you looked into it, anyway 🙏
- I've tried with Dev Mode on to do the same (as before it was giving the "You are not allowed"), and now instead of deleting the account, it doesn't do anything, refreshes back into the Account overview, but doesn't give any error or anything. Is this the expected behavior with the last changes?

**Comment by dmallory42** [comment]
Thanks for testing, @daquinons  - as mentioned on Slack I've created a follow-up issue to handle that edge case: https://github.com/Automattic/woocommerce-payments/issues/7803.
 
> This happened with two different JN configurations
> 
>     I selected WooCommerce, WooPayments, and this branch from the beta tool
>     I selected WooCommerce, this branch from the beta tool, but did not select WooPayments
> 
> This is likely just a JN issue but I wanted to flag it anyway just in case there's something in there to be fixed.
> 
> If I skip the store profiler at the point of error I arrive at the Woo home, and can click through to WooPayments and can go through the setup to get into dev mode without any issues. Once in dev mode I see our new prompts and they click through to the first step of Payments Connect without any issues.

@bradyv this is a quirk of using live branches/beta tool. Essentially what's happening is the profiler is trying to install the latest version of the WooPayments plugin, which is likely causing some sort of conflict or failure with the live branch version installed on the JN site. Thanks for flagging it though!

**Comment by bradyv** [comment]
Ahhh okay. In that case I believe this is good to go from a design pov!

**Pull request merged**